CDCgov/prime-reportstream
# Engagement/angela/11293/sender receiver fhir cli

This PR ...
Adds the ability to use both sender and receiver schemas on the `./prime fhirdata` CLI command. Note: If you had previously been using `-s` in a command, you will want to use `-r` instead now, as the original functionality was representative of a receiver schema.

Test Steps:
1. Run `./prime fhirdata --input-file src/testIntegration/resources/datatests/FHIR_to_HL7/sample_SR_1_20230302-0001.fhir --output-format FHIR --output-file junk/test/SR-FHIR-NY-X.fhir`
2. Verify output tells you that you must specify a schema
3. Run `./prime fhirdata --input-file src/testIntegration/resources/datatests/FHIR_to_HL7/sample_SR_1_20230302-0001.fhir --output-format HL7 --output-file junk/test/SR-FHIR-NY-X.hl7`
4. Verify output tells you that you must specify a receiver schema (Receiver schema is required to convert to HL7)
5. Run `./prime fhirdata --input-file src/testIntegration/resources/datatests/FHIR_to_HL7/sample_SR_1_20230302-0001.fhir --output-format HL7 -s metadata/fhir_transforms/senders/default-sender-transform.yml --output-file junk/test/SR-FHIR-NY-X.hl7`
6. Verify output tells you that you must specify a receiver schema
7. Run `./prime fhirdata --input-file src/testIntegration/resources/datatests/FHIR_to_HL7/sample_SR_1_20230302-0001.fhir --output-format HL7 -r metadata/hl7_mapping/receivers/STLTs/NY/NY-receiver-transform.yml --output-file junk/test/SR-FHIR-NY-R.hl7`
8. Verify that SR-FHIR-NY-R.hl7 was produced, is an HL7 file, and that MSH-4 is `CDC PRIME REPORTSTREAM^10DRPTSTRM^CLIA`
9. Run `./prime fhirdata --input-file src/testIntegration/resources/datatests/FHIR_to_HL7/sample_SR_1_20230302-0001.fhir --output-format HL7 -s metadata/fhir_transforms/senders/default-sender-transform.yml -r metadata/hl7_mapping/receivers/STLTs/NY/NY-receiver-transform.yml --output-file junk/test/SR-FHIR-NY.hl7`
10. Verify that SR-FHIR-NY.hl7 was produced, is an HL7 file, that MSH-3 is `CDC PRIME - Atlanta^2.16.840.1.114222.4.1.237821^ISO` and that MSH-4 is `CDC PRIME REPORTSTREAM^10DRPTSTRM^CLIA` 
11. Run `./prime fhirdata --input-file src/testIntegration/resources/datatests/FHIR_to_HL7/sample_SR_1_20230302-0001.fhir --output-format FHIR -s metadata/fhir_transforms/senders/default-sender-transform.yml --output-file junk/test/SR-FHIR-NY-S.fhir`
12. Verify that SR-FHIR-NY-S.fhir was produced, is a FHIR file, and that the name of the source in MessageHeader resourceType is `"CDC PRIME - Atlanta: 2.16.840.1.114222.4.1.237821"`
13. Run `./prime fhirdata --input-file src/testIntegration/resources/datatests/FHIR_to_HL7/sample_SR_1_20230302-0001.fhir --output-format FHIR -r metadata/fhir_transforms/receivers/fhir-transform-sample.yml --output-file junk/test/SR-FHIR-NY-R.fhir`
14. Verify that SR-FHIR-NY-R.fhir was produced, is a FHIR file, and that the name text in the Patient resourceType is `TestValue`
15. Run `./prime fhirdata --input-file src/testIntegration/resources/datatests/FHIR_to_HL7/sample_SR_1_20230302-0001.fhir --output-format FHIR -r metadata/fhir_transforms/receivers/fhir-transform-sample.yml -s metadata/fhir_transforms/senders/default-sender-transform.yml --output-file junk/test/SR-FHIR-NY.fhir`
16. Verify that SR-FHIR-NY.fhir was produced, is a FHIR file, that the name of the source in MessageHeader resourceType is `"CDC PRIME - Atlanta: 2.16.840.1.114222.4.1.237821"`, and that the name text in the Patient resourceType is `TestValue`
17. Run `./prime fhirdata --input-file src/testIntegration/resources/datatests/HL7_to_FHIR/sample_orm_20230809-001.hl7 --output-format FHIR -r metadata/fhir_transforms/receivers/fhir-transform-sample.yml -s metadata/fhir_transforms/senders/default-sender-transform.yml --output-file junk/test/SR-HL7-NY.fhir`
18. Verify that SR-HL7-NY.fhir was produced, is a FHIR file, that the name of the source in MessageHeader resourceType is `"CDC PRIME - Atlanta: 2.16.840.1.114222.4.1.237821"`, and that the name text in the Patient resourceType is `TestValue`
19. Run `./prime fhirdata --input-file src/testIntegration/resources/datatests/HL7_to_FHIR/sample_orm_20230809-001.hl7 --output-format FHIR --output-file junk/test/SR-HL7-NY-X.fhir`
20. Verify that SR-HL7-NY-X.fhir was produced and is a FHIR file (Note: This is an input HL7 file to our internal format)
21. Run `./prime fhirdata --input-file src/testIntegration/resources/datatests/HL7_to_FHIR/sample_orm_20230809-001.hl7 --output-format FHIR -r metadata/fhir_transforms/receivers/fhir-transform-sample.yml --output-file junk/test/SR-HL7-NY-R.fhir`
22. Verify that SR-HL7-NY-R.fhir was produced, is a FHIR file, and that the name text in the Patient resourceType is `TestValue`
23. Run `./prime fhirdata --input-file src/testIntegration/resources/datatests/HL7_to_FHIR/sample_orm_20230809-001.hl7 --output-format FHIR -s metadata/fhir_transforms/senders/default-sender-transform.yml --output-file junk/test/SR-HL7-NY-S.fhir`
24. Verify that SR-HL7-NY-S.fhir was produced, is a FHIR file, that the name of the source in MessageHeader resourceType is `"CDC PRIME - Atlanta: 2.16.840.1.114222.4.1.237821"`
25. Run `./prime fhirdata --input-file src/testIntegration/resources/datatests/HL7_to_FHIR_to_HL7/elims_1_72_3029202646_5532_NoPII.hl7 --output-format HL7 -r metadata/hl7_mapping/receivers/STLTs/NY/NY-receiver-transform.yml -s metadata/fhir_transforms/senders/default-sender-transform.yml --output-file junk/test/SR-HL7-NY.hl7`
26. Verify that SR-HL7-NY.hl7 was produced, is an HL7 file, that MSH-3 is `CDC PRIME - Atlanta^2.16.840.1.114222.4.1.237821^ISO` and that MSH-4 is `CDC PRIME REPORTSTREAM^10DRPTSTRM^CLIA` 
27. Run `./prime fhirdata --input-file src/testIntegration/resources/datatests/HL7_to_FHIR_to_HL7/elims_1_72_3029202646_5532_NoPII.hl7 --output-format HL7 --output-file junk/test/SR-HL7-NY-X.hl7`
28. Verify output tells you that you must specify a receiver schema (Receiver schema is required to conver to HL7)
29. Run `./prime fhirdata --input-file src/testIntegration/resources/datatests/HL7_to_FHIR_to_HL7/elims_1_72_3029202646_5532_NoPII.hl7 --output-format HL7 -s metadata/fhir_transforms/senders/default-sender-transform.yml --output-file junk/test/SR-HL7-NY-S.hl7`
30. Verify output tells you that you must specify a receiver schema (Receiver schema is required to convert to HL7)
31. Run `./prime fhirdata --input-file src/testIntegration/resources/datatests/HL7_to_FHIR_to_HL7/elims_1_72_3029202646_5532_NoPII.hl7 --output-format HL7 -r metadata/hl7_mapping/receivers/STLTs/NY/NY-receiver-transform.yml --output-file junk/test/SR-HL7-NY-R.hl7`
32. Verify that SR-HL7-NY-R.hl7 was produced, is an HL7 file, and that MSH-4 is `CDC PRIME REPORTSTREAM^10DRPTSTRM^CLIA`

## Changes
-  Changed `./prime fhirdata` flag `-s, --schema` to `-r, --receiver-schema`
- Added `./prime fhirdata` flag `-s, --sender-schema`
- Updated method names for clarity
- Updated methods to make use of both sender and receiver schemas if present

## Checklist

### Testing

### Process

## Linked Issues
- Fixes #11293
